### DISCUSSION TITLE
Compute perplexity of TransformerLM

### DISCUSSION THREAD

**Discussion by PabloPeso** [speechbrain/speechbrain#1664] [Q&A]
Hi,

I was wondering if there are Speech Brain built-in tools to compute the perplexity of a LM model?

I can see there is a `ngram_perplexity()` function but I am not sure how use it, specially with the transformer-based LM.

Thanks,

Pablo Peso.

**Comment by TParcollet** [comment]
Hi, I am not sure that we have such a functionality. We decided not to add it because it simply is linked to the cross-entropy observed on the model.

**Comment by PabloPeso** [comment]
Hi,

Yes, that's right, it is related to the cross-entropy. However, in case I just have two pre-trained LM models, how could I easily compare them? I guess your point is that I could use the evaluate function used in the `train.py` script, for example
https://github.com/speechbrain/speechbrain/blob/a9b373a9fef3fdfc4e3b4a41478f7b283866186f/recipes/LibriSpeech/LM/train.py#L209
and compare their test loss, is that right?

Thanks for your help.

Pablo Peso.

**Comment by TParcollet** [comment]
Hi that's right ! Or you can develop a function to obtain the perplexity and we could add it to the toolkit as well.

**Comment by PabloPeso** [comment]
Computing the perplexity should be straight forward from the current loss defined in the config file (`nll_loss`). Following https://huggingface.co/docs/transformers/perplexity the perplexity should be the exponential of the negative log-likelihood loss.

On the other hand, what would be the easiest way to load a checkpoint into the lm_brain object (created from LM class inherit from sb.core.Brain class)? I was checking the `load_checkpoint` function from Checkpoints.py, but it is expecting the CKPT.yaml which is sometimes inaccessible (like for example if we want to compute the perplexity of https://huggingface.co/speechbrain/asr-transformer-transformerlm-librispeech/blob/main/lm.ckpt).

Thanks

**Comment by TParcollet** [comment]
If it's to evaluate the model (call the evaluate on it) then you need to use the standard training recipe + checkpointer. The CKPT.yaml is given on the GDRive folder: https://github.com/speechbrain/speechbrain/tree/develop/recipes/LibriSpeech/ASR/transformer

HuggingFace is only here for inference :-)
